pkg/engine/resolve/inputtemplate.go

```diff
@@ -37,23 +36,28 @@ type InputTemplate struct {
 
 var setTemplateOutputNull = errors.New("set to null")
 
-func (i *InputTemplate) Render(ctx *Context, data []byte, preparedInput *fastbuffer.FastBuffer) (err error) {
-	undefinedVariables := make([]string, 0)
+func (i *InputTemplate) Render(ctx *Context, data []byte, preparedInput *fastbuffer.FastBuffer) error {
+	var undefinedVariables []string
 
-	for j := range i.Segments {
-		switch i.Segments[j].SegmentType {
+	for _, segment := range i.Segments {
```

Comment:
The reason we've used index access here is to not create copies of the segments, which your rewrite now does. Effects might be minimal, but not zero. Do you have a good reason to use copies here?

I'm not opposed to copies, just want to understand why this refactor was necessary.

Reply:
It is the other way around. With the range using the index you're making a new copy every time you read from the slice:

- 44:`switch i.Segments[j].SegmentType`
- 48:`switch i.Segments[j].VariableKind `
- 54:`i.Segments[j].VariableSourcePath`
- 56:`i.Segments[j].VariableKind`

(I'm not counting copies that pass  i.Segments[j] by value to another function because we cannot elide those)

Ranging with a value instead we make a copy at the start of the loop and then we use that across the whole iteration, creating 3 less copies.

In fact, the generated assembly seems to support this ('-' is ranging with value, '+' is ranging with index):

```diff
-github.com/wundergraph/graphql-go-tools/pkg/engine/resolve.(*InputTemplate).Render STEXT size=1504 args=0x30 locals=0x188 funcid=0x0 align=0x0
-       0x0000 00000 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   TEXT    github.com/wundergraph/graphql-go-tools/pkg/engine/resolve.(*InputTemplate).Render(SB), ABIInternal, $400-48
+github.com/wundergraph/graphql-go-tools/pkg/engine/resolve.(*InputTemplate).Render STEXT size=1680 args=0x30 locals=0x1e8 funcid=0x0 align=0x0
+       0x0000 00000 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   TEXT    github.com/wundergraph/graphql-go-tools/pkg/engine/resolve.(*InputTemplate).Render(SB), ABIInternal, $496-48
        0x0000 00000 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   MOVD    16(g), R16
        0x0004 00004 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   PCDATA  $0, $-2
-       0x0004 00004 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   SUB     $272, RSP, R17
+       0x0004 00004 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   SUB     $368, RSP, R17
        0x0008 00008 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   CMP     R16, R17
-       0x000c 00012 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   BLS     1436
+       0x000c 00012 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   BLS     1612
        0x0010 00016 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   PCDATA  $0, $-1
-       0x0010 00016 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   SUB     $400, RSP, R20
+       0x0010 00016 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   SUB     $496, RSP, R20
        0x0014 00020 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   STP     (R29, R30), -8(R20)
        0x0018 00024 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   PCDATA  $0, $-2
        0x0018 00024 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   MOVD    R20, RSP
        0x001c 00028 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   PCDATA  $0, $-1
        0x001c 00028 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   SUB     $8, RSP, R29
-       0x0020 00032 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   FUNCDATA        $0, gclocals·IL4UPPBcICQL/TGWCkBKmw==(SB)
-       0x0020 00032 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   FUNCDATA        $1, gclocals·sbbtSWTKH8TzPzMpN5aMwg==(SB)
+       0x0020 00032 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   FUNCDATA        $0, gclocals·DTMxmu7RUpolfv2Y2eSf3w==(SB)
+       0x0020 00032 (/Users/fiam/Source/WunderGraph/graphql-go-tools/pkg/engine/resolve/inputtemplate.go:40)   FUNCDATA        $1, gclocals·6Nlgdw1ZQ7pFG8AxIyzd6w==(SB)
```

Notice how with these changes the code size decreases from 1680 to 1504 and the locals siz decreases from 0x1e8 to 0x188. That's 96 bytes less, matching the 3 less copies of `InputTemplate` that we counted by hand, which is 32 bytes.